// midas/dist/types.js
'use strict';

exports.__esModule = true;
var atRule = exports.atRule = 'at-rule';
var atRuleName = exports.atRuleName = 'at-rule-name';
var atRuleKeyword = exports.atRuleKeyword = 'at-rule-keyword';
var attribute = exports.attribute = 'attribute';
var attributeBrace = exports.attributeBrace = 'attribute-brace';
var attributeInsensitive = exports.attributeInsensitive = 'attribute-insensitive';
var attributeName = exports.attributeName = 'attribute-name';
var attributeOperator = exports.attributeOperator = 'attribute-operator';
var attributeValue = exports.attributeValue = 'attribute-value';
var brace = exports.brace = 'brace';
var className = exports.className = 'class';
var colon = exports.colon = 'colon';
var combinator = exports.combinator = 'combinator';
var comment = exports.comment = 'comment';
var func = exports.func = 'function';
var funcName = exports.funcName = 'function-name';
var hexColor = exports.hexColor = 'hex-color';
var hexValue = exports.hexValue = 'hex-value';
var id = exports.id = 'id';
var important = exports.important = 'important';
var number = exports.number = 'number';
var parenthesis = exports.parenthesis = 'parenthesis';
var property = exports.property = 'property';
var pseudo = exports.pseudo = 'pseudo';
var selector = exports.selector = 'selector';
var semicolon = exports.semicolon = 'semicolon';
var string = exports.string = 'string';
var tag = exports.tag = 'tag';
var universal = exports.universal = 'universal';
var value = exports.value = 'value';
var vendorPrefix = exports.vendorPrefix = 'vendor-prefix';
var word = exports.word = 'word';
